Add UCM for PinePhone

tomfitzhenry opened the pull request:
Fixes https://github.com/alsa-project/alsa-ucm-conf/issues/124.

This PR adds support for the https://wiki.pine64.org/index.php/PinePhone 

This PR contains:
* This UCM config copied from https://gitlab.com/pine64-org/pine64-alsa-ucm/, which was developed by many developers ( https://gitlab.com/pine64-org/pine64-alsa-ucm/-/merge_requests/1 ) and is BSD-3-Clause licensed.
* Fix paths to be relative to alsa-ucm-conf root.
* Some refactoring of the UCM config, incl. removing duplicate directives, and moving volume directives to BootSequence.
* Updating the routing/playback/boost volumes, in response to comments on this PR.

I have tested:

* HiFi
    * Record audio via the headset.
    * Record audio via the internal mic.
    * Play audio on the headphones.
    * Play audio on the speaker.
    * Play audio on the internal speaker.
* Voice Call
    * Speak on the headset.
    * Speak on the internal mic.
    * Listen on the headphones.
    * Listen on the speaker. This induces an echo, and given the audio routing is entirely on chip, and the chip doesn't advertise echo cancellation, I don't see how this is avoidable.
    * Listen on the internal speaker.

See also:
* `alsa-info`: https://alsa-project.org/db/?f=5e331f5ef10bd167b3bd7640bfcf0e879100f80d
* `pa-info`: [pa-info.txt](https://github.com/alsa-project/alsa-ucm-conf/files/8015083/pa-info.txt)
* https://gitlab.com/pine64-org/pine64-alsa-ucm/-/issues/3
* Schematics/code
    * The device-tree configuration has had a unique audio card name since 5.14: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi?h=v5.14#n436
    * http://files.pine64.org/doc/PinePhone/PinePhone%20v1.2%20Released%20Schematic.pdf
    * https://xnux.eu/devices/feature/audio-pp.html
    * https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi
    * https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/allwinner/sun50i-a64.dtsi
    * https://github.com/torvalds/linux/blob/master/sound/soc/sunxi/sun50i-codec-analog.c
    * https://files.pine64.org/doc/datasheet/pine64/Allwinner_A64_User_Manual_V1.0.pdf
    * https://static.chipdip.ru/lib/126/DOC001126567.pdf

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,146 @@
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
+	]
+
+        DisableSequence [
```
This uses spaces, which is inconsistent with the rest of the file (tabs).

tomfitzhenry replied:
Done.

a-wai commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,146 @@
+		cset "name='Mic1 Capture Switch' off"
+		cset "name='Mic1 Boost Volume' 7"
+		cset "name='Mic2 Capture Switch' off"
+		cset "name='Mic2 Boost Volume' 7"
```
```suggestion
```

tomfitzhenry replied:
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1113/diffs.

From that merge request, it's not clear why this was introduced.

tomfitzhenry replied:
See https://github.com/alsa-project/alsa-ucm-conf/pull/134/files#r805993961 for discussion.

tomfitzhenry replied:
I've applied your suggestion.

a-wai commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,151 @@
+		cset "name='Mic1 Boost Volume' 0"
+		cset "name='Mic1 Playback Volume' 7"
+		cset "name='Mic2 Capture Switch' off"
+		cset "name='Mic2 Boost Volume' 0"
+		cset "name='Mic2 Playback Volume' 7"
```
```suggestion
```

tomfitzhenry replied:
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1652 and is well discussed. I will test, and discuss with the pmos folks on this.

tomfitzhenry replied:
SGTM, per https://github.com/alsa-project/alsa-ucm-conf/pull/134/files#r800623320.

tomfitzhenry replied:
I have tested this with a headset, and observe that with any `Mic2 Boost Volume` that is >0 induces non-trivial noise.

The best I can do is:
* Mic2 Playback Volume at 0 dB
* Mic2 Boost Volume at 0 dB
* ADC Capture Volume at 24 dB (max)
* ADC Gain at 6dB (max)

That comes out loud and clear.

tomfitzhenry replied:
I've applied your suggestion.

tomfitzhenry replied:
The default Mic2 Boost Volume seems to be 33 dB, not 0 dB. Maybe we will need to set `Mic2 Boost Volume` to 0. Is this better done in the kernel?

tomfitzhenry replied:
Fixed by explicitly setting `Mic2 Boost Volume` to 0 dB.

a-wai commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,151 @@
+		cset "name='Mixer Capture Switch' off"
+		cset "name='Mixer Reversed Capture Switch' off"
+		cset "name='ADC Capture Volume' 160"
+		cset "name='ADC Gain Capture Volume' 7"
```
```suggestion
```

tomfitzhenry replied:
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1113/diffs.

From that merge request, it's not clear why this was introduced.

tomfitzhenry replied:
I lean towards committing this suggestion, on the basis that we should prefer boosting the analog signal over boosting the digital signal. In https://github.com/alsa-project/alsa-ucm-conf/pull/134/files#r800623320, the analog signal boost will occur.

I will test this.

tomfitzhenry replied:
Upon testing ( `arecord -D 'sysdefault:CARD=PinePhone' -c2 -f S16_LE  foo.wav; aplay -D 'sysdefault:CARD=PinePhone' foo.wav`) I observe that 33dB boost is acceptably loud, and 42dB boost induces noise.

I'll commit this suggestion.

a-wai reviewed (changes requested):
Thanks for taking care of this!

I've added a few comments (mostly regarding mic boost and ADC gain) based on my testing (`HiFi` only for the latter, couldn't test your version of `VoiceCall` but instead suggested the values I use on a daily basis with Mobian).

tomfitzhenry commented:
> I've added a few comments (mostly regarding mic boost and ADC gain) based on my testing (HiFi only for the latter, couldn't test your version of VoiceCall but instead suggested the values I use on a daily basis with Mobian).

Thanks for your comprehensive review!

Your comments make sense. For each comment I will search through the commits of https://gitlab.com/pine64-org/pine64-alsa-ucm to see the directives have any justification.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,146 @@
+		cset "name='Mixer Capture Switch' off"
+		cset "name='Mixer Reversed Capture Switch' off"
+		cset "name='ADC Capture Volume' 160"
+		cset "name='ADC Gain Capture Volume' 7"
```
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1113/diffs.

From that merge request, it's not clear why this was introduced.

tomfitzhenry replied:
I lean towards committing this suggestion, on the basis that we should prefer boosting the analog signal over boosting the digital signal. In https://github.com/alsa-project/alsa-ucm-conf/pull/134/files#r800623320, the analog signal boost will occur.

I will test this.

tomfitzhenry replied:
Upon testing ( `arecord -D 'sysdefault:CARD=PinePhone' -c2 -f S16_LE  foo.wav; aplay -D 'sysdefault:CARD=PinePhone' foo.wav`) I observe that default 0db gain is acceptably loud, and the extra 6dB boost induces noise.

I'll commit this suggestion.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,146 @@
+
+		cset "name='Line In Capture Switch' off"
+		cset "name='Mic1 Capture Switch' off"
+		cset "name='Mic1 Boost Volume' 7"
```
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1113/diffs.

From that merge request, it's not clear why this was introduced.

tomfitzhenry replied:
Upon testing ( `arecord -D 'sysdefault:CARD=PinePhone' -c2 -f S16_LE  foo.wav; aplay -D 'sysdefault:CARD=PinePhone' foo.wav`) I observe that 33dB boost is acceptably loud, and 42dB boost induces noise.

I'll commit this suggestion.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,151 @@
+		cset "name='AIF2 ADC Mixer ADC Capture Switch' on"
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF2 DAC Rev Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
```
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/938/diffs and I think it's safe to remove.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,146 @@
+		cset "name='AIF2 ADC Mixer ADC Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF2 DAC Rev Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
```
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/938/diffs and I think it's safe to remove.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,146 @@
+	EnableSequence [
+		cset "name='AIF1 DA0 Stereo Playback Route' Stereo"
+		cset "name='Headphone Playback Switch' on"
+		cset "name='Headphone Playback Volume' 70%"
```
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/2182 and is well discussed. I will test, and discuss with the pmos folk on this.

smaeul replied:
This control is the gain for the headphone analog power amplifier. I don't think the units here are dB relative to input (it's an amplifier after all), but dB relative to maximum amplification. So this gain should be as small as possible to minimize noise.

a-wai replied:
Ah right, my "0dB obsession" fooled me again! So 70% should be fine here, my bad.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,151 @@
+
+		cset "name='Line In Capture Switch' off"
+		cset "name='Mic1 Capture Switch' off"
+		cset "name='Mic1 Boost Volume' 0"
+		cset "name='Mic1 Playback Volume' 7"
```
This was introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1652 and is well discussed. I will test, and discuss with the pmos folks on this.

tomfitzhenry replied:
I lean towards committing this suggestion, on the basis that we should prefer boosting the analog signal over boosting the digital signal. The change in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1652 appears to have been an over-correction of https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1113/diffs.

I will test this.

tomfitzhenry replied:
Upon testing ( `arecord -D 'sysdefault:CARD=PinePhone' -c2 -f S16_LE  foo.wav; aplay -D 'sysdefault:CARD=PinePhone' foo.wav`) I observe that the default are acceptably loud.

I'll commit this suggestion.

tomfitzhenry commented:
The remaining comments are non-trivial changes.

I will test the Pinephone audio with the comments having applied.

@a-wai, I see Mobian is already using the comments as you've described ( https://gitlab.com/mobian1/packages/alsa-ucm-conf/-/blob/mobian/debian/patches/0001-ucm2-add-PinePhone-configuration.patch ) which gives confidence that they'll work.

For each profile, I will test:
* Headphones
* Internal speaker
* Internal earpiece
* Internal mic
* Mic via audio jack

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,144 @@
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF2 DAC Rev Capture Switch' off"
+	]
+
+	DisableSequence [
+	]
+
+	Value {
+	}
+}
```
Can these empty sections be removed?

smaeul commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,144 @@
+SectionDevice."Earpiece" {
+	Comment "Internal Earpiece"
+	EnableSequence [
+		cset "name='AIF1 DA0 Stereo Playback Route' Mix Mono"
```
The "Speaker" and "Earpiece" sections don't conflict with "Headphones". So enabling both "Speaker" and "Headphones" would give you mix mono audio to the headphones. I think a nice solution here is, at the HiFi profile level:
 * Enable "DAC Playback Switch"
 * Enable "DAC Reversed Playback Switch"
 * Set "AIF1 DA0 Stereo Playback Route" to "Stereo"
 * Set "Headphone Source Playback Route" to "DAC"
 * Set "Earpiece Source Playback Route" to "Left Mixer"

Then, the headphones will always be stereo, because they bypass the analog mixer. At the same time, since both DAC channels will go into both sides of the analog mixer, the earpiece and speaker will get mix mono.

tomfitzhenry replied:
If "DAC Reversed Playback Switch" is enabled on the HiFi profile, and an XMPP video/audio chat is active, then AFAICT the output would be fed into the input, as if it came from the mic.

AFAICT, an XMPP video/audio chat wouldn't want to use the Voice Call profile, since the Voice Call profile sets "PlaybackRate 8000".

As an alternative, should we just set that "Speaker" and "Earpiece" each conflict with "Headphones". I think this matches user expectations too.

smaeul replied:
"DAC Playback Switch" and "DAC Reversed Playback Switch" control the exact same path from the DACs to the analog Output Mixer, except that the left and right channels are swapped. They are the two pairs of "m"s overlaying the rightward-pointing arrows just to the right of the green "DACL" and "DACR" boxes in this diagram: https://xnux.eu/devices/feature/audio-controls.svg.

The "Reversed" here refers to the stereo channels, not to the direction.

(The only loopback paths in the codec are controlled by the "Mixer (Reversed) Capture Switch"; see the leftward-pointing arrows at the bottom of the diagram. Those are always off in both profiles.)

tomfitzhenry replied:
Thanks for the explanation, that makes sense.

I like this approach because it means our routing can be only: routes don't change, only switches do. That makes this config easier to understand.

I have updated this PR accordingly: https://github.com/alsa-project/alsa-ucm-conf/pull/134/commits/fb6e94983e69b902c66306214a4cfab3beabb7fa.

smaeul commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,144 @@
+		cset "name='Mic2 Playback Switch' off"
+		cset "name='AIF1 DA0 Playback Volume' 160"
+		cset "name='AIF3 ADC Source Capture Route' None"
+		cset "name='AIF2 DAC Source Playback Route' None"
```
"None" is not a valid value for this control. When not using AIF3, this should be set to "AIF2". See here: https://github.com/torvalds/linux/blob/master/sound/soc/sunxi/sun8i-codec.c#L844-L855

smaeul commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,149 @@
+		cset "name='DAC Playback Switch' on"
+		cset "name='DAC Playback Volume' 160"
+		cset "name='ADC Digital DAC Playback Switch' off"
+		cset "name='AIF1 Slot 0 Digital DAC Playback Switch' on"
```
AIF1 Slot 0 is the CPU audio interface. Is it intentional that this is enabled during a call?

tomfitzhenry replied:
Perhaps folks want to record their calls? Or maybe GNOME Calls uses the VoiceCall even for SIP?

What's the disadvantage of leaving this enabled?

smaeul replied:
I suppose there isn't any downside. This won't cause any additional hardware to be powered up, since the capture path must be enabled anyway for recording.

One note though on call recording: you will need to enable "AIF2 Digital ADC Capture Switch" in order to capture audio from the modem. ("AIF1 Data Digital ADC Capture Switch" only gets you audio from the analog codec, i.e. the microphone).

tomfitzhenry replied:
Done.

tomfitzhenry commented:
For posterity, here's the resources I'm reading to learn more about ALSA/UCM/Pinephone:
* http://files.pine64.org/doc/PinePhone/PinePhone%20v1.2%20Released%20Schematic.pdf
* https://xnux.eu/devices/feature/audio-pp.html
* https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi
* https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/allwinner/sun50i-a64.dtsi
* https://github.com/torvalds/linux/blob/master/sound/soc/sunxi/sun50i-codec-analog.c
* https://files.pine64.org/doc/datasheet/pine64/Allwinner_A64_User_Manual_V1.0.pdf

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,149 @@
+		cset "name='AIF1 DA0 Playback Volume' 160"
+		cset "name='AIF2 DAC Playback Volume' 160"
+		cset "name='AIF3 ADC Source Capture Route' None"
+		cset "name='AIF2 DAC Source Playback Route' None"
```
`None` is invalid, per https://github.com/alsa-project/alsa-ucm-conf/pull/134/files#r802213404.

tomfitzhenry replied:
Introduced in https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1614, and looks to pre-date smaeul's work.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,143 @@
+	EnableSequence [
+		cset "name='AIF1 DA0 Stereo Playback Route' Stereo"
+		cset "name='Headphone Playback Switch' on"
+		cset "name='Headphone Playback Volume' 100%"
```
Should this be 70%, like as decided in https://github.com/alsa-project/alsa-ucm-conf/pull/134/files#r800616927 ?

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,143 @@
+	]
+
+	Value {
+		PlaybackRate 8000
```
This is the only directive that meaningfully differs between VoiceCall and HiFi. Perhaps the similarity can be deduplicated (e.g. move to PinePhone.conf somehow?)

tomfitzhenry commented:
@a-wai, @smaeul: Thanks for your reviews, I've addressed your comments.

From here I will:
* do a final test
* add some documentation, and maybe an updated https://xnux.eu/devices/feature/modem-audio-route.svg
* seek review from the ALSA UCM folk, in particular around how we could de-duplicate the config between the two Usecases.

I'd appreciate if you found the time to do a final review and/or test.

tomfitzhenry commented:
Two issues:
* commit ec8e6ced80711b4824c36d059b9d2cc5682bbebe "PinePhone: HiFi: Match VoiceCall's Mic1 CapturePriority" appears to have broken capture selection when selecting a capture device during a GNOME Calls call. I will remove this commit.
* The headset mic is too noisy. I've unresolved the relevant comment.

tomfitzhenry commented:
> add some documentation, and maybe an updated https://xnux.eu/devices/feature/modem-audio-route.svg

@megous, do you know which license (if any) this diagram is available under? I'd like to modify it and include it as part of this PR.

philmmanjaro commented:
If I compare this UCM config to the one by Mobian, I still see some differences: [mobian.diff.txt](https://github.com/alsa-project/alsa-ucm-conf/files/8172438/mobian.diff.txt) Also people reported this: The microphone on phone calls is super sensitive and in the settings I could not lower to any effect. This made calls nearly impossible. Installing pavucontrol (pulse audio volume control) and lowering mic input to about 15% returned calls to normal. If you use Sound Record on Phosh you have some high pitch background noise during the recording.

philmmanjaro commented:
Also you can't mute the mic in **gnome-calls**. On a pinephone-pro with their UCM config it seems to work. See also here: https://gitlab.com/mobian1/callaudiod/-/merge_requests/10

philmmanjaro commented:
Seems it also depends if you use **pulseaudio** or **pipewire**. At least **Sound Recorder** may behave differently.

philmmanjaro commented:
Differences from the Pine64 UCM profile: [p64u.diff.txt](https://github.com/alsa-project/alsa-ucm-conf/files/8172909/p64u.diff.txt)

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,135 @@
+	Comment "Internal speaker"
+	EnableSequence [
+		cset "name='Line Out Playback Switch' on"
+		cset "name='Line Out Playback Volume' 100%"
```
"It is not recommended to set the fixed values for the volume settings to the Enable / Disable sequences for verbs or devices, if the device exports the hardware volume (MixerElem or Volume/Switch values). The default volume settings should be set in the BootSequence." https://www.alsa-project.org/alsa-doc/alsa-lib/group__ucm__conf.html

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,135 @@
+	Comment "Internal Earpiece"
+	EnableSequence [
+		cset "name='Earpiece Playback Switch' on"
+		cset "name='Earpiece Playback Volume' 100%"
```
"It is not recommended to set the fixed values for the volume settings to the Enable / Disable sequences for verbs or devices, if the device exports the hardware volume (MixerElem or Volume/Switch values). The default volume settings should be set in the BootSequence." https://www.alsa-project.org/alsa-doc/alsa-lib/group__ucm__conf.html

tomfitzhenry replied:
Done.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,135 @@
+	Comment "Headphones"
+	EnableSequence [
+		cset "name='Headphone Playback Switch' on"
+		cset "name='Headphone Playback Volume' 70%"
```
"It is not recommended to set the fixed values for the volume settings to the Enable / Disable sequences for verbs or devices, if the device exports the hardware volume (MixerElem or Volume/Switch values). The default volume settings should be set in the BootSequence." https://www.alsa-project.org/alsa-doc/alsa-lib/group__ucm__conf.html

tomfitzhenry replied:
Done.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,141 @@
+	Comment "Internal speaker"
+	EnableSequence [
+		cset "name='Line Out Playback Switch' on"
+		cset "name='Line Out Playback Volume' 100%"
```
"It is not recommended to set the fixed values for the volume settings to the Enable / Disable sequences for verbs or devices, if the device exports the hardware volume (MixerElem or Volume/Switch values). The default volume settings should be set in the BootSequence." https://www.alsa-project.org/alsa-doc/alsa-lib/group__ucm__conf.html

tomfitzhenry replied:
Done.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,141 @@
+	Comment "Internal Earpiece"
+	EnableSequence [
+		cset "name='Earpiece Playback Switch' on"
+		cset "name='Earpiece Playback Volume' 100%"
```
"It is not recommended to set the fixed values for the volume settings to the Enable / Disable sequences for verbs or devices, if the device exports the hardware volume (MixerElem or Volume/Switch values). The default volume settings should be set in the BootSequence." https://www.alsa-project.org/alsa-doc/alsa-lib/group__ucm__conf.html

tomfitzhenry replied:
Done.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,141 @@
+	Comment "Headphones"
+	EnableSequence [
+		cset "name='Headphone Playback Switch' on"
+		cset "name='Headphone Playback Volume' 70%"
```
"It is not recommended to set the fixed values for the volume settings to the Enable / Disable sequences for verbs or devices, if the device exports the hardware volume (MixerElem or Volume/Switch values). The default volume settings should be set in the BootSequence." https://www.alsa-project.org/alsa-doc/alsa-lib/group__ucm__conf.html

tomfitzhenry replied:
Done.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,135 @@
+		cset "name='Line In Capture Switch' off"
+		cset "name='Mic1 Capture Switch' off"
+		cset "name='Mic2 Capture Switch' off"
+		cset "name='Mic2 Boost Volume' 0"
```
I found a headset where Mic2 Boost Volume = 0 results in no audio. (Maybe it always did?)

tomfitzhenry replied:
This was a faulty headset.

philmmanjaro commented:
I normally use pine64-ucm but also explore what upstream or others do. Seems at least we got some useful feedback from users.

bhush9 commented:
Regarding mute,

> Ah, good spot. I wasn't aware of this usecase. I'm trying to prioritise upstreaming, rather than support new usecases. Do you know if this has ever worked for the Pinephone?

@tomfitzhenry I found fix for this, If you add `CaptureMixerElem "ADC"` in the capture nodes, hardware mute seems to work fine, which makes muting audio call work perfectly (at least verified using parecord and there it seems to work fine).

Can you add this in your UCM?

https://gitlab.com/pine64-org/pine64-alsa-ucm/-/merge_requests/7

tomfitzhenry committed:
Add UCM for PinePhone

The DTS configuration has had a unique audio card name since 5.14:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi?h=v5.14#n436d

This UCM config is derived from https://gitlab.com/pine64-org/pine64-alsa-ucm/, which is BSD-3-Clause licensed.

Signed-off-by: Tom Fitzhenry <tom@tom-fitzhenry.me.uk>
Co-authored-by: Arnaud Ferraris <arnaud.ferraris@collabora.com>
Co-authored-by: Samuel Holland <samuel@sholland.org>

tomfitzhenry commented:
> Can you add this in your UCM?
>
> https://gitlab.com/pine64-org/pine64-alsa-ucm/-/merge_requests/7

Per https://gitlab.com/pine64-org/pine64-alsa-ucm/-/merge_requests/7#note_881261109 , I'll hold off including that PR for now.

tomfitzhenry commented:
I've tested this PR, and it's ready for review. I have updated the PR description with the testing done.

a-wai commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,129 @@
+		cset "name='ADC Digital DAC Playback Switch' off"
+		cset "name='AIF1 Slot 0 Digital DAC Playback Switch' on"
+		cset "name='AIF2 Digital DAC Playback Switch' off"
+		cset "name='DAC Reversed Playback Switch' on"
```
May I assume this routes `DACR` to `MIXL` (and `DACL` to `MIXR`) so `MIX*` are both identival and effectively a mono mix of the DAC outputs?

smaeul replied:
Yes, that is correct.

tomfitzhenry replied:
I will document this.

a-wai replied:
Awesome, thanks! (that's a clever use of this output mixer)

a-wai reviewed (changes requested):
Sorry it took so long for a new round of review.

I'm still thinking the volume/gain controls should be reworked (i.e. more similar to what distros currently use), but otherwise it looks good, with some nice additions.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/PinePhone.conf:
```diff
@@ -0,0 +1,28 @@
+BootSequence [
+	cset "name='AIF1 DA0 Playback Volume' 100%"
+	cset "name='AIF2 DAC Playback Volume' 100%"
+	cset "name='DAC Playback Volume' 100%"
```
My bad, I converted `160` to `100%`, mistakenly believed it was a no-op. Fixed.

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/PinePhone.conf:
```diff
@@ -0,0 +1,28 @@
+	cset "name='Mic2 Boost Volume' 1"
+
+	cset "name='ADC Capture Volume' 100%"
+	cset "name='AIF1 AD0 Capture Volume' 100%"
```
> Same comment here: those controls are just fine using 160 / 0dB, but make audio unusable when set to 100% (even with Mic1 Boost Volume set to 0)

My bad, I converted `160` to `100%`, mistakenly believed it was a no-op. Fixed.

> Mic1 Boost: not set here, assuming 0 (0dB)

The default is 3. I will confirm this.

Mic2 (headset) appears to need less boost on the headsets I have tried. I will re-confirm.

> As a general rule of thumb, I believe you should start with all volume controls on 0dB (except for the headphones output for the reasons discussed here already) and add small touches only where needed, starting with mic boosts.

Sounds good, this was the intention. The conversion of `160` to `100%` broke this, however. Now fixed.

a-wai replied:
> My bad, I converted `160` to `100%`, mistakenly believed it was a no-op. Fixed.

Oh ok, I understand, those things can be misleading indeed.

> > Mic1 Boost: not set here, assuming 0 (0dB)
> 
> The default is 3. I will confirm this.

I believe you're right, it should default to whatever the hardware sets, which is 33db (`3`) in this case. It should be all good then.

> Mic2 (headset) appears to need less boost on the headsets I have tried. I will re-confirm.

Finding the "right" value here might be tricky as there can be big variations between headsets. Maybe stick with the default? (although I have no strong opinion on that one, and my headset's mic is broken so I can't really test).

> Sounds good, this was the intention. The conversion of `160` to `100%` broke this, however. Now fixed.

Got it, thanks!

tomfitzhenry commented on ucm2/Allwinner/A64/PinePhone/HiFi.conf:
```diff
@@ -0,0 +1,129 @@
+		cset "name='AIF3 ADC Source Capture Route' None"
+		cset "name='AIF2 DAC Source Playback Route' AIF2"
+		# AIF2 (Modem) is mono.
+		cset "name='AIF2 DAC Stereo Playback Route' Mix Mono"
```
I'm keen to do this. I tried it with FixedBootSequence, but this appeared to occur before the driver had loaded. I will try BootSequence for the switches, and read https://www.alsa-project.org/alsa-doc/alsa-lib/group__ucm__conf.html to understand how this interacts with users who choose to override some volume/switches.

a-wai replied:
`FixedBootSequence` would be nice indeed, I wonder why it didn't work when you tested. In this case `BootSequence` might hopefully work just as well, but that's to be thoroughly tested.

ollieparanoid commented:
Thanks everybody who worked on this!

@tomfitzhenry: you marked this as draft, what's the reason for that? Can we help with getting this finished up?

ollieparanoid commented:
@tomfitzhenry, again thanks a bunch for all your work on this! I've tested it on postmarketOS edge. The headset mic seems to be routed wrong, everything else works fine. I've tested with a headset with a four-contact (TRRS) plug.

| Test | Result |
| --- | --- |
|  HiFi |  |
| * Record audio via the headset. | NOK (1) |
| * Record audio via the internal mic. | OK |
| * Play audio on the headphones. | OK (2) |
| * Play audio on the speaker. | OK |
| * Play audio on the internal speaker. | OK |
| Voice Call | |
| * Speak on the headset. | NOK (3) |
| * Speak on the internal mic. | OK |
| * Listen on the headphones. | OK |
| * Listen on the speaker. | OK (4) | 
| * Listen on the internal speaker. | OK |

Notes:

1: tried with gnome sound recorder, the line just stays completely flat (wrong routing? see also 3)

2: it's a bit louder than the default we currently have in postmarketOS. We had turned the default down because as it was at max, one could hear noise in a quiet room. If I concentrate, I think I can hear some noise very quietly now, but it's so quiet it's fine in my opinion.

3: routing seems to be wrong here. I tried a call and could hear me talking into the headset on the pinephone itself instead of on the other phone. So for some reason it gets routed to the headset output?

4: I can reproduce the echo you described above.


The config looks fine from a quick glance (though I don't really have experience with reading/writing alsa ucm configs). In my opinion, once the headset mic bug is fixed, this is ready to be merged.

perexg commented:
The single commit is preferred for the initial work, Thank you.

ollieparanoid commented:
> > I'm away-from-Pinephone at the moment, so won't be able to re-test the NOK tests for a few weeks.
> 
> I've re-tested these, and I observe that they work for me.

Nice!

> @ollieparanoid, if you could re-test these cases in light of the comments I left in [#134 (comment)](https://github.com/alsa-project/alsa-ucm-conf/pull/134#issuecomment-1220231690) that'd be appreciated.

Yes, sorry for not getting back to that earlier. I should be able to test it over the weekend.

ollieparanoid commented:
Tested again:

| Test 					| Result |
| --- 					| --- |
|  HiFi 				|  |
| * Record audio via the headset.	| ??? (1) |
| * Record audio via the internal mic.	| OK |
| * Play audio on the headphones.	| OK |
| * Play audio on the speaker.		| OK |
| * Play audio on the internal speaker. | OK |
| Voice Call				| |
| * Speak on the headset.		| ??? (1) |
| * Speak on the internal mic.		| OK |
| * Listen on the headphones.		| OK |
| * Listen on the speaker.		| OK | 
| * Listen on the internal speaker.	| OK |

(1) So everything is fine for me except for the headset mic. I've been testing this for more than an hour now, and in that time I have gotten record audio via the headset working only once and by simply rebooting a lot of times and trying it again. When it worked, I had the headset plugged. But I wasn't able to reproduce it, every other time I've tried it just wasn't able to record anything from the mic again. I also tried a different headset in case this one was broken, but that wasn't the problem.

Also I found that this is not a regression with the config from this PR. When I run the alsa-ucm-conf we currently use in postmarketOS, I have the same problem (didn't realize/test this last time, typically I don't use the headset mic much). I'm not sure why this happens, might be a hardware defect with my particular device or anything else.

> Can you confirm that /var/lib/alsa/asound.state is non-existent at the time of this PR's config having been loaded? Otherwise lingering ALSA commands from previous config might be conflicting with this PR.

Yes, I can confirm that there was no asound.state file during the testing I did now, and probably also not when I tested last time.

> Also, could you try raising the Mic2 Boost Volume to 3 (which can be done via alsamixer)? This is the default, which I'm inclined to have this PR use (as has been suggested by reviewers)? Is the headset mic audible then?

I'm using pulseaudio and alsamixer didn't show the option there. I also didn't find it in pauvcontrol. I guess I would need to change the audio device in alsamixer? (if you want me to test it again, to what?) - but I'm not sure if it would change much given that most of the time the mic doesn't work at all (boost should only increase the volume gain, right?). And the one time it was working, I could hear it clearly, it wasn't very silent.

>> 3: routing seems to be wrong here. I tried a call and could hear me talking into the headset on the pinephone itself instead of on the other phone. So for some reason it gets routed to the headset output?
>
> Could this be sound captured from your other phone's mic? If I perform this test with my other phone close by, I also hear my voice on the headset, but if I mute my other phone, I no longer hear my voice on the headset.

Oh yes that was the reason! When I muted the other phone I couldn't hear anything while speaking to the headset mic.

So now that we know that the failures I had earlier are not regressions from changing to this config, and everything else works fine, I would recommend that this gets merged.

a-wai commented:
Finally took some time to test, sorry about the delay.

In my case everything works as expected (including headset mic, but we're not using the same kernel), although I had to manually reset the earpiece volume to 100%; this is likely due to me having set it to 0 a while ago (this is my test device, so I do all sorts of fiddling on that one) and therefore not caused by this PR.

I therefore grant this PR my "Approved" stamp ;)

perexg commented on ucm2/Allwinner/A64/PinePhone/PinePhone.conf:
```diff
@@ -0,0 +1,65 @@
+}
+
+FixedBootSequence [
+	# Switch playback off.
```
I don't think that this initialization scheme is quite correct. The verb EnableSequence or global SectionDefaults sequence should disable the controlled devices  - https://www.alsa-project.org/alsa-doc/alsa-lib/group__ucm__conf.html - to put the hardware to a good initial state before any device from the verb is used. Move the "turn mixer paths off" commands to SectionDefaults for multiple verbs. It will be executed before any verb is activated.

The other controls may be moved to the standard BootSequence, too. But if they are not supposed (or dangerous) to be modified by user later, you can keep them here. Note that alsactl should restore the previous state of all ALSA driver controls on boot.

perexg commented on ucm2/Allwinner/A64/PinePhone/VoiceCall.conf:
```diff
@@ -0,0 +1,105 @@
+		PlaybackPCM "hw:${CardId},0"
+	}
+}
+SectionDevice."Earpiece" {
```
Missing new-line.

tomfitzhenry committed:
fixup: add missing newlines

tomfitzhenry committed:
fixup: remove superfluous Playback/CaptureChannels 2

tomfitzhenry committed:
fixup: use SectionDefaults

tomfitzhenry committed:
fixup: define PlaybackMixerElem

perexg reviewed (changes requested):
SectionDefaults should be in PinePhone.conf (like BootSequence). It is shared by all profiles (and used when profile changes).

tomfitzhenry committed:
fixup: factor out SectionDefaults to PinePhone.conf

philmmanjaro commented:
For whatever reason with this new config there is no sound at all on the Pinephone when using Manjaro.

philmmanjaro commented:
Hmm, interesting @ollieparanoid and @tomfitzhenry. There is something missing when used on bleeding edge distros. Issue with [PMOS](https://gitlab.com/postmarketOS/pmaports/-/issues/2115) and with [Manjaro](https://gitlab.manjaro.org/manjaro-arm/issues/pinephone/phosh/-/issues/282). What is the best way to debug this?

nullr0ute commented:
> Hmm, interesting @ollieparanoid and @tomfitzhenry. There is something missing when used on bleeding edge distros. Issue with [PMOS](https://gitlab.com/postmarketOS/pmaports/-/issues/2115) and with [Manjaro](https://gitlab.manjaro.org/manjaro-arm/issues/pinephone/phosh/-/issues/282). What is the best way to debug this?

I wonder if there's a difference between their kernel forks and pure upstream linus kernel.

philmmanjaro commented:
Dowstream distros starting to revert these changes to known working ones. Example: https://raw.githubusercontent.com/dreemurrs-embedded/Pine64-Arch/master/PKGBUILDS/pine64/alsa-ucm-conf/0001-Revert-to-Mobian-UCM.patch

a-wai commented:
Just tested on a fresh Mobian install with both PulseAudio (v16.1) and PipeWire (v0.3.71) + WirePlumber (v0.4.14) and everything seems alright, except for the earpiece: anything playing on the right channel will be dropped, only the left channel actually plays back audio; I suspect that might cause issues with the modem audio, so I've created #324 to address this.
